## Learning Objectives

- **Apply broadcasting rules** for operations between arrays of different shapes  
- **Optimize code performance** using vectorization instead of loops  
- **Execute mathematical operations** (arithmetic, linear algebra, statistics)  
- **Generate random data** for simulations and statistical analysis  
- **Solve data science problems** using NumPy's computational power  
- **Convert between NumPy and Pandas** for optimal workflow efficiency 
- **Perform vectorized computations** that are 10–100x faster than Python loops  

In [2]:
import numpy as np
import pandas as pd

##  Vectorization in NumPy

### Introduction to Vectorization
Vectorization is the process of applying operations to entire arrays or matrices simultaneously rather than iterating over individual elements using loops. This approach allows NumPy to utilize highly optimized C libraries for faster computation.

![](https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcTQ_9hGutgPqNO9OCLy-vDyJub4iLpgThGZ9A&s)

Benefits of Vectorization:

* **Performance**: Significantly faster than using for-loops in Python (often 10-100x speedup)
* **Conciseness**: Shorter and more readable code
* **Efficiency**: Reduces the overhead of Python loops and leverages lower-level optimizations
* **Parallelization**: Many vectorized operations can utilize multiple CPU cores
* **Memory Efficiency**: Better cache utilization and reduced memory overhead

### Why Vectorization Matters

**Traditional Python Loop Approach:**
```python
# Slow: Python loop with individual operations
result = []
for i in range(len(array1)):
    for j in range(len(array2)):
        result.append(array1[i] * array2[j] + some_value)
```

**Vectorized NumPy Approach:**
```python
# Fast: Single vectorized operation
result = array1[:, np.newaxis] * array2 + some_value
```

The vectorized approach is faster because:
1. **Compiled C code**: NumPy operations use optimized C/Fortran libraries
2. **Reduced Python overhead**: Fewer Python function calls and type checks
3. **SIMD instructions**: Can use Single Instruction, Multiple Data operations
4. **Better memory access patterns**: More cache-friendly data access

###  Understanding Vectorized Operations with Examples

In [3]:

# Create two arrays
arr1 = np.array([1, 2, 3, 4, 5])
arr2 = np.array([10, 20, 30, 40, 50])

# Element-wise addition (vectorized operation)
sum_arr = arr1 + arr2
print("Sum Array:", sum_arr)

Sum Array: [11 22 33 44 55]


### Types of Vectorized Operations

NumPy supports several categories of vectorized operations:

#### 1. **Element-wise Operations (Universal Functions - ufuncs)**
These operations apply a function to each element of an array:

In [5]:
# Basic arithmetic operations (all vectorized)
arr = np.array([1, 4, 9, 16, 25])

print("Original array:", arr)
print("Square root:", np.sqrt(arr))
print("Natural log:", np.log(arr))
print("Sine:", np.sin(arr))
print("Power of 2:", np.power(arr, 2))

# Comparison operations
print("\nComparison operations:")
print("Greater than 10:", arr > 10)
print("Equal to 9:", arr == 9)
print("Between 5 and 20:", (arr >= 5) & (arr <= 20))

Original array: [ 1  4  9 16 25]
Square root: [1. 2. 3. 4. 5.]
Natural log: [0.         1.38629436 2.19722458 2.77258872 3.21887582]
Sine: [ 0.84147098 -0.7568025   0.41211849 -0.28790332 -0.13235175]
Power of 2: [  1  16  81 256 625]

Comparison operations:
Greater than 10: [False False False  True  True]
Equal to 9: [False False  True False False]
Between 5 and 20: [False False  True  True False]


#### 2. **Aggregation Operations**
These reduce arrays to scalar values or smaller arrays:

In [6]:
# 2D array for aggregation examples
matrix = np.array([[1, 2, 3, 4],
                   [5, 6, 7, 8],
                   [9, 10, 11, 12]])

print("Matrix:")
print(matrix)

# Aggregations across entire array
print(f"\nSum of all elements: {np.sum(matrix)}")
print(f"Mean: {np.mean(matrix)}")
print(f"Standard deviation: {np.std(matrix)}")
print(f"Min: {np.min(matrix)}, Max: {np.max(matrix)}")

# Aggregations along specific axes
print(f"\nSum along axis 0 (columns): {np.sum(matrix, axis=0)}")
print(f"Sum along axis 1 (rows): {np.sum(matrix, axis=1)}")
print(f"Mean along axis 0: {np.mean(matrix, axis=0)}")
print(f"Max along axis 1: {np.max(matrix, axis=1)}")

Matrix:
[[ 1  2  3  4]
 [ 5  6  7  8]
 [ 9 10 11 12]]

Sum of all elements: 78
Mean: 6.5
Standard deviation: 3.452052529534663
Min: 1, Max: 12

Sum along axis 0 (columns): [15 18 21 24]
Sum along axis 1 (rows): [10 26 42]
Mean along axis 0: [5. 6. 7. 8.]
Max along axis 1: [ 4  8 12]


#### 3. **Boolean Operations and Fancy Indexing**
Vectorized selection and filtering operations:

In [7]:
# Boolean indexing - vectorized filtering
data = np.array([1, 15, 8, 23, 4, 16, 42, 3, 19, 7])

# Find elements meeting multiple conditions (vectorized)
mask = (data > 5) & (data < 20)
filtered_data = data[mask]
print(f"Original data: {data}")
print(f"Elements between 5 and 20: {filtered_data}")

# Using np.where for conditional replacement (vectorized)
# Replace values > 15 with -1, others with original value
result = np.where(data > 15, -1, data)
print(f"After conditional replacement: {result}")

# Fancy indexing - select multiple elements at once
indices = [0, 2, 4, 7]
selected = data[indices]
print(f"Elements at indices {indices}: {selected}")

# Count how many elements meet condition (vectorized)
count_large = np.sum(data > 10)
print(f"Count of elements > 10: {count_large}")

Original data: [ 1 15  8 23  4 16 42  3 19  7]
Elements between 5 and 20: [15  8 16 19  7]
After conditional replacement: [ 1 15  8 -1  4 -1 -1  3 -1  7]
Elements at indices [0, 2, 4, 7]: [1 8 4 3]
Count of elements > 10: 5


###  Performance Comparison: NumPy Vectorization Vs Python Loops

Understanding the performance benefits of vectorization is crucial for writing efficient scientific computing code. Let's compare different scenarios to see when and why vectorization matters.

#### Scenario 1: Basic Mathematical Operations
`np.dot` is a vectorized operation in NumPy that performs matrix multiplication or dot product between arrays. It efficiently computes the element-wise multiplications and then sums them up. To better understand its efficiency, let's first implement the dot product using for loops, and then compare its performance with `np.dot` to see the benefits of vectorization.

In [8]:
import time

# Function to calculate dot product using for loops
def dot_product_loops(arr1, arr2):
    result = 0
    for i in range(len(arr1)):
        result += arr1[i] * arr2[i]
    return result

# Create sample arrays of different sizes for comparison
sizes = [1000, 10000, 100000, 1000000]

print("Performance Comparison: Loops vs NumPy Vectorization")
print("=" * 60)

for size in sizes:
    # Create random arrays
    arr1 = np.random.rand(size)
    arr2 = np.random.rand(size)
    
    # Measure time for the loop-based implementation
    start_time = time.time()
    loop_result = dot_product_loops(arr1, arr2)
    loop_time = time.time() - start_time
    
    # Measure time for np.dot
    start_time = time.time()
    numpy_result = np.dot(arr1, arr2)
    numpy_time = time.time() - start_time
    
    # Calculate speedup
    speedup = loop_time / numpy_time if numpy_time > 0 else float('inf')
    
    print(f"\nArray size: {size:,}")
    print(f"Loop result: {loop_result:.5f}, Time: {loop_time:.5f}s")
    print(f"NumPy result: {numpy_result:.5f}, Time: {numpy_time:.5f}s")
    print(f"Speedup: {speedup:.1f}x faster")
    
    

print(f"\n{'='*60}")
print("Key Observations:")
print("• NumPy becomes dramatically faster as array size increases")
print("• Speedup can reach 50-100x for large arrays")
print("• NumPy overhead is minimal for small arrays but pays off quickly")

Performance Comparison: Loops vs NumPy Vectorization

Array size: 1,000
Loop result: 234.28341, Time: 0.00000s
NumPy result: 234.28341, Time: 0.00000s
Speedup: infx faster

Array size: 10,000
Loop result: 2524.58396, Time: 0.00000s
NumPy result: 2524.58396, Time: 0.02460s
Speedup: 0.0x faster

Array size: 100,000
Loop result: 24892.29841, Time: 0.02218s
NumPy result: 24892.29841, Time: 0.00104s
Speedup: 21.4x faster

Array size: 1,000,000
Loop result: 249763.97969, Time: 0.20611s
NumPy result: 249763.97969, Time: 0.00111s
Speedup: 185.6x faster

Key Observations:
• NumPy becomes dramatically faster as array size increases
• Speedup can reach 50-100x for large arrays
• NumPy overhead is minimal for small arrays but pays off quickly


#### Scenario 2: Complex Mathematical Operations
Let's compare more complex operations that show even greater performance differences:

In [9]:
# Complex mathematical operation: polynomial evaluation
# f(x) = 3x³ + 2x² - 5x + 1

def polynomial_loops(x_values):
    """Evaluate polynomial using loops"""
    results = []
    for x in x_values:
        result = 3*x**3 + 2*x**2 - 5*x + 1
        results.append(result)
    return results

def polynomial_vectorized(x_values):
    """Evaluate polynomial using vectorized operations"""
    return 3*x_values**3 + 2*x_values**2 - 5*x_values + 1

# Test with different sizes
size = 500000
x_data = np.random.uniform(-10, 10, size)

print("Complex Operations: Polynomial Evaluation")
print("f(x) = 3x³ + 2x² - 5x + 1")
print("-" * 40)

# Loop version
start_time = time.time()
loop_results = polynomial_loops(x_data)
loop_time = time.time() - start_time

# Vectorized version
start_time = time.time()
vector_results = polynomial_vectorized(x_data)
vector_time = time.time() - start_time

speedup = loop_time / vector_time

print(f"Array size: {size:,}")
print(f"Loop time: {loop_time:.4f}s")
print(f"Vectorized time: {vector_time:.4f}s")
print(f"Speedup: {speedup:.1f}x")

# Verify results match
print(f"Results match: {np.allclose(loop_results, vector_results)}")

Complex Operations: Polynomial Evaluation
f(x) = 3x³ + 2x² - 5x + 1
----------------------------------------
Array size: 500,000
Loop time: 0.2560s
Vectorized time: 0.0129s
Speedup: 19.9x
Results match: True


#### Performance Summary & Key Takeaways

The performance comparisons clearly demonstrate that **vectorized NumPy operations significantly outperform equivalent Python loops**:

### Broadcasting: The Heart of NumPy Vectorization

Broadcasting is NumPy's powerful mechanism that allows operations between arrays of different shapes without explicitly reshaping them. It's the foundation that makes vectorization so flexible and intuitive.

#### Broadcasting Rules

NumPy broadcasting follows these rules:

1. **Start from the trailing dimension** and work backwards
2. **Dimensions are compatible** if:
   - They are equal, OR
   - One of them is 1, OR  
   - One of them doesn't exist (missing dimension)
3. **Missing dimensions** are assumed to be size 1

#### Visual Guide to Broadcasting

```
Array 1: (3, 4)     [[1, 2, 3, 4],      Array 2: (4,)    [10, 20, 30, 40]
                     [5, 6, 7, 8],                         ↓ broadcasts to
                     [9,10,11,12]]                         [[10, 20, 30, 40],
                                                            [10, 20, 30, 40], 
                                                            [10, 20, 30, 40]]
Result: (3, 4)      Element-wise addition produces (3, 4) result
```

Let's explore different broadcasting scenarios:

In [10]:
# Comprehensive Broadcasting Examples

print("=== BROADCASTING EXAMPLES ===\n")

# Example 1: 2D array + 1D array (most common case)
arr_2d = np.array([[1, 2, 3, 4], 
                   [5, 6, 7, 8], 
                   [9, 10, 11, 12]])

arr_1d = np.array([10, 20, 30, 40])

print("Example 1: (3,4) + (4,) broadcasting")
print(f"2D array shape: {arr_2d.shape}")
print(f"1D array shape: {arr_1d.shape}")
result_1 = arr_2d + arr_1d
print(f"Result shape: {result_1.shape}")
print("Result:")
print(result_1)
print()

# Example 2: Broadcasting with different orientations
arr_col = np.array([[100], 
                    [200], 
                    [300]])  # Column vector (3,1)

print("Example 2: (3,4) + (3,1) broadcasting")
print(f"2D array shape: {arr_2d.shape}")
print(f"Column vector shape: {arr_col.shape}")
result_2 = arr_2d + arr_col
print(f"Result shape: {result_2.shape}")
print("Result:")
print(result_2)
print()

# Example 3: Scalar broadcasting (always works)
scalar = 1000
print("Example 3: (3,4) + scalar broadcasting")
result_3 = arr_2d + scalar
print(f"Result shape: {result_3.shape}")
print("Result:")
print(result_3)
print()

# Example 4: 3D broadcasting
arr_3d = np.random.randint(1, 10, (2, 3, 4))
arr_2d_broadcast = np.array([[1, 2, 3, 4]])  # (1, 4)

print("Example 4: (2,3,4) + (1,4) broadcasting")
print(f"3D array shape: {arr_3d.shape}")
print(f"2D array shape: {arr_2d_broadcast.shape}")
result_4 = arr_3d + arr_2d_broadcast
print(f"Result shape: {result_4.shape}")
print("Broadcasting successful!")
print()

=== BROADCASTING EXAMPLES ===

Example 1: (3,4) + (4,) broadcasting
2D array shape: (3, 4)
1D array shape: (4,)
Result shape: (3, 4)
Result:
[[11 22 33 44]
 [15 26 37 48]
 [19 30 41 52]]

Example 2: (3,4) + (3,1) broadcasting
2D array shape: (3, 4)
Column vector shape: (3, 1)
Result shape: (3, 4)
Result:
[[101 102 103 104]
 [205 206 207 208]
 [309 310 311 312]]

Example 3: (3,4) + scalar broadcasting
Result shape: (3, 4)
Result:
[[1001 1002 1003 1004]
 [1005 1006 1007 1008]
 [1009 1010 1011 1012]]

Example 4: (2,3,4) + (1,4) broadcasting
3D array shape: (2, 3, 4)
2D array shape: (1, 4)
Result shape: (2, 3, 4)
Broadcasting successful!



In [11]:
# Broadcasting Rules Demonstration

def check_broadcast_compatibility(shape1, shape2):
    """
    Check if two shapes are compatible for broadcasting
    """
    # Pad shorter shape with 1s on the left
    ndim1, ndim2 = len(shape1), len(shape2)
    max_ndim = max(ndim1, ndim2)
    
    shape1_padded = [1] * (max_ndim - ndim1) + list(shape1)
    shape2_padded = [1] * (max_ndim - ndim2) + list(shape2)
    
    result_shape = []
    compatible = True
    
    for i in range(max_ndim):
        dim1, dim2 = shape1_padded[i], shape2_padded[i]
        if dim1 == dim2:
            result_shape.append(dim1)
        elif dim1 == 1:
            result_shape.append(dim2)
        elif dim2 == 1:
            result_shape.append(dim1)
        else:
            compatible = False
            break
    
    return compatible, tuple(result_shape) if compatible else None

# Test different shape combinations
test_cases = [
    ((3, 4), (4,)),      # Compatible
    ((3, 4), (3, 1)),    # Compatible  
    ((3, 4), (2,)),      # Incompatible
    ((2, 3, 4), (4,)),   # Compatible
    ((2, 3, 4), (3, 4)), # Compatible
    ((2, 3, 4), (2, 1, 4)), # Compatible
    ((3, 4), (2, 3)),    # Incompatible
]

print("Broadcasting Compatibility Check:")
print("-" * 50)
for shape1, shape2 in test_cases:
    compatible, result_shape = check_broadcast_compatibility(shape1, shape2)
    status = "✅ Compatible" if compatible else "❌ Incompatible"
    if compatible:
        print(f"{str(shape1):>10} + {str(shape2):>10} → {str(result_shape):>12} {status}")
    else:
        print(f"{str(shape1):>10} + {str(shape2):>10} → {'N/A':>12} {status}")

Broadcasting Compatibility Check:
--------------------------------------------------
    (3, 4) +       (4,) →       (3, 4) ✅ Compatible
    (3, 4) +     (3, 1) →       (3, 4) ✅ Compatible
    (3, 4) +       (2,) →          N/A ❌ Incompatible
 (2, 3, 4) +       (4,) →    (2, 3, 4) ✅ Compatible
 (2, 3, 4) +     (3, 4) →    (2, 3, 4) ✅ Compatible
 (2, 3, 4) +  (2, 1, 4) →    (2, 3, 4) ✅ Compatible
    (3, 4) +     (2, 3) →          N/A ❌ Incompatible


#### Real-World Broadcasting Applications

Broadcasting enables elegant solutions to common data science problems:

In [12]:
# Real-world broadcasting applications

# Application 1: Data Normalization (Z-score standardization)
print("=== APPLICATION 1: DATA NORMALIZATION ===")
# Sample data: test scores for 100 students across 5 subjects
np.random.seed(42)
scores = np.random.normal(75, 15, (100, 5))  # 100 students, 5 subjects
subjects = ['Math', 'Science', 'English', 'History', 'Art']

print(f"Original data shape: {scores.shape}")
print(f"Sample scores (first 5 students):")
for i in range(5):
    print(f"Student {i+1}: {scores[i].round(1)}")

# Calculate mean and std for each subject (broadcasting will handle the rest)
subject_means = np.mean(scores, axis=0)  # Shape: (5,)
subject_stds = np.std(scores, axis=0)    # Shape: (5,)

print(f"\nSubject means: {subject_means.round(1)}")
print(f"Subject stds: {subject_stds.round(1)}")

# Normalize using broadcasting: (100,5) - (5,) / (5,) → (100,5)
normalized_scores = (scores - subject_means) / subject_stds

print(f"\nNormalized data shape: {normalized_scores.shape}")
print(f"Normalized means (should be ~0): {np.mean(normalized_scores, axis=0).round(3)}")
print(f"Normalized stds (should be ~1): {np.std(normalized_scores, axis=0).round(3)}")
print()

# Application 2: Distance Matrix Calculation  
print("=== APPLICATION 2: DISTANCE MATRIX ===")
# Calculate pairwise distances between points using broadcasting
points = np.array([[1, 2], [3, 4], [5, 6], [7, 8]])  # 4 points in 2D
n_points = points.shape[0]

print(f"Points shape: {points.shape}")
print("Points:")
for i, point in enumerate(points):
    print(f"  Point {i}: ({point[0]}, {point[1]})")

# Using broadcasting to calculate all pairwise distances efficiently
# points[:, np.newaxis, :] has shape (4, 1, 2)  
# points[np.newaxis, :, :] has shape (1, 4, 2)
# Broadcasting gives us (4, 4, 2) - all pairwise differences
differences = points[:, np.newaxis, :] - points[np.newaxis, :, :]
distances = np.sqrt(np.sum(differences**2, axis=2))

print(f"\nDistance matrix shape: {distances.shape}")
print("Distance matrix:")
print(distances.round(2))
print()

# Application 3: Image Processing with Broadcasting
print("=== APPLICATION 3: IMAGE PROCESSING ===")
# Simulate a grayscale image and apply different brightness adjustments to RGB channels
np.random.seed(42)
image_gray = np.random.randint(0, 256, (100, 100))  # 100x100 grayscale image

# Create RGB image by broadcasting
rgb_multipliers = np.array([1.0, 0.8, 0.6])  # R, G, B channel multipliers
# Reshape for broadcasting: (100, 100, 1) * (3,) → (100, 100, 3)
image_rgb = image_gray[:, :, np.newaxis] * rgb_multipliers

print(f"Original grayscale image shape: {image_gray.shape}")
print(f"RGB multipliers: {rgb_multipliers}")
print(f"Final RGB image shape: {image_rgb.shape}")
print(f"RGB image dtype: {image_rgb.dtype}")

# Show some pixel values
print(f"\nSample pixel transformation:")
print(f"Original gray value: {image_gray[50, 50]}")
print(f"RGB values: {image_rgb[50, 50].round(1)}")
print()

print("💡 Key Broadcasting Benefits:")
print("  • No explicit loops required")
print("  • Memory efficient (no array copying)")  
print("  • Highly optimized C implementation")
print("  • Readable and concise code")
print("  • Scales to any dimension")

=== APPLICATION 1: DATA NORMALIZATION ===
Original data shape: (100, 5)
Sample scores (first 5 students):
Student 1: [82.5 72.9 84.7 97.8 71.5]
Student 2: [71.5 98.7 86.5 68.  83.1]
Student 3: [68.  68.  78.6 46.3 49.1]
Student 4: [66.6 59.8 79.7 61.4 53.8]
Student 5: [97.  71.6 76.  53.6 66.8]

Subject means: [74.3 76.2 73.6 76.9 74.5]
Subject stds: [13.4 14.7 14.7 14.3 16. ]

Normalized data shape: (100, 5)
Normalized means (should be ~0): [ 0.  0. -0. -0. -0.]
Normalized stds (should be ~1): [1. 1. 1. 1. 1.]

=== APPLICATION 2: DISTANCE MATRIX ===
Points shape: (4, 2)
Points:
  Point 0: (1, 2)
  Point 1: (3, 4)
  Point 2: (5, 6)
  Point 3: (7, 8)

Distance matrix shape: (4, 4)
Distance matrix:
[[0.   2.83 5.66 8.49]
 [2.83 0.   2.83 5.66]
 [5.66 2.83 0.   2.83]
 [8.49 5.66 2.83 0.  ]]

=== APPLICATION 3: IMAGE PROCESSING ===
Original grayscale image shape: (100, 100)
RGB multipliers: [1.  0.8 0.6]
Final RGB image shape: (100, 100, 3)
RGB image dtype: float64

Sample pixel transforma

In [13]:
arr2 = np.array([[1, 2, 3, 4], 
                 [5, 6, 7, 8], 
                 [9, 1, 2, 3]])

In [14]:
arr4 = np.array([4, 5, 6, 7])

In [15]:
arr2 + arr4

array([[ 5,  7,  9, 11],
       [ 9, 11, 13, 15],
       [13,  6,  8, 10]])

When the expression `arr2 + arr4` is evaluated, `arr4` (which has the shape `(4,)`) is replicated three times to match the shape `(3, 4)` of `arr2`. Numpy performs the replication without actually creating three copies of the smaller dimension array, thus improving performance and using lower memory.

<img src="https://jakevdp.github.io/PythonDataScienceHandbook/figures/02.05-broadcasting.png" width="360">

Broadcasting only works if one of the arrays can be replicated to match the other array's shape.

* Dimensions of size 1 will broadcast (as if the value was repeated). 
* Otherwise, the dimension must have the same shape. 

In [16]:
arr5 = np.array([7, 8])

In [19]:
# uncomment the code to see the error
# arr2 + arr5

In the above example, even if `arr5` is replicated three times, it will not match the shape of `arr2`. Hence `arr2 + arr5` cannot be evaluated successfully. See the [broadcasting documentation](https://numpy.org/doc/stable/user/basics.broadcasting.html) to learn more about it.

### Matrix Multiplication in NumPy with Vectorization

Matrix multiplication is one of the most common and computationally intensive operations in numerical computing and deep learning. NumPy offers efficient and highly optimized methods for performing matrix multiplication, which leverage vectorization to handle large matrices quickly and accurately.

**Note that**: NumPy matrix operations follow the standard rules of linear algebra, so it’s important to ensure that the shapes of the matrices are compatible. If they are not, consider reshaping the matrices before performing multiplication

There are two commonly methods for matrix multiplication

#### Method 1: Matrix Multiplication Using `np.dot()`

In [20]:
# Define two 2D arrays (matrices)
matrix1 = np.array([[1, 2, 3], 
                    [4, 5, 6]])
matrix2 = np.array([[7, 8], 
                    [9, 10], 
                    [11, 12]])

# Matrix multiplication using np.dot
result_dot = np.dot(matrix1, matrix2)
print("Matrix Multiplication using np.dot:\n", result_dot)

# Another way to perform np.dot for matrix multiplication
result_dot2 = matrix1.dot(matrix2)
print("\nMatrix Multiplication using dot method:\n", result_dot2)


Matrix Multiplication using np.dot:
 [[ 58  64]
 [139 154]]

Matrix Multiplication using dot method:
 [[ 58  64]
 [139 154]]


#### Method 2: Matrix Multiplication using `np.matmul()` or `@`

In [21]:
# Matrix multiplication using np.matmul or @ operator
result_matmul = np.matmul(matrix1, matrix2)
result_operator = matrix1 @ matrix2
print("\nMatrix Multiplication using np.matmul:\n", result_matmul)
print("\nMatrix Multiplication using @ operator:\n", result_operator)



Matrix Multiplication using np.matmul:
 [[ 58  64]
 [139 154]]

Matrix Multiplication using @ operator:
 [[ 58  64]
 [139 154]]


Note that `*` operator in numpy is element-wise multiplication

In [24]:
# using * operator for element-wise multiplication, please uncomment the code below to run
# element_wise = matrix1 * matrix2
# print("\nElement-wise Multiplication:\n", element_wise)

In [25]:
# reshape the array for element-wise multiplication
matrix2_reshaped = matrix2.reshape(2, 3)
element_wise = matrix1 * matrix2_reshaped
print("\nElement-wise Multiplication after reshaping:\n", element_wise)


Element-wise Multiplication after reshaping:
 [[ 7 16 27]
 [40 55 72]]


##  Converting Between NumPy Arrays and Pandas DataFrames

**Seamless data conversion** between NumPy and pandas is essential for efficient data science workflows. Each library excels in different areas, and knowing when and how to convert between them maximizes your analytical power.

###  Why Conversion Matters

**NumPy** and **pandas** complement each other perfectly:

| **Aspect** | **NumPy Arrays** | **Pandas DataFrames** |
|------------|------------------|----------------------|
| **Strength** | Mathematical computations | Data manipulation & analysis |
| **Speed** | ⚡ Faster numerical operations | 🔍 Easier data exploration |
| **Memory** | 💾 More memory efficient | 🏷️ Rich metadata (labels, dtypes) |
| **Use Cases** | Linear algebra, statistics, ML algorithms | Data cleaning, grouping, merging |
| **Indexing** | Integer-based only | Label-based + integer-based |
| **Missing Data** | Limited NaN support | 🛡️ Robust NaN handling |



**Typical Workflow**:
1. **Load data** with pandas (CSV, Excel, databases)
2. **Clean and prepare** with pandas (filtering, grouping)
3. **Convert to NumPy** for intensive computations (ML, statistics)
4. **Convert back to pandas** for results analysis and visualization

###  NumPy Array → Pandas DataFrame

**When to convert**: You have computational results in NumPy arrays and need pandas' data manipulation capabilities.

**Common scenarios**:

-  Adding meaningful labels to numerical results
-  Exporting to CSV, Excel, or databases
-  Preparing data for visualization libraries
-  Using pandas' powerful filtering and grouping

####  Basic Conversion Methods



In [26]:
# 📊 COMPREHENSIVE ARRAY → DATAFRAME CONVERSION
import numpy as np
import pandas as pd

print("🔄 NUMPY TO PANDAS CONVERSION")
print("=" * 40)

🔄 NUMPY TO PANDAS CONVERSION


In [27]:
# Original NumPy array
scores = np.array([
    [85, 92, 78, 95],
    [88, 76, 91, 82],
    [95, 89, 84, 90]
])

students = ['Alice', 'Bob', 'Carol']
subjects = ['Math', 'Science', 'English', 'History']

print("Original NumPy array:")
print(scores)
print(f"Shape: {scores.shape}")

Original NumPy array:
[[85 92 78 95]
 [88 76 91 82]
 [95 89 84 90]]
Shape: (3, 4)


In [28]:
# Method 1: Basic conversion (auto index & column labels: 0..n-1)
df_basic = pd.DataFrame(scores)
print("\n📋 Basic DataFrame (auto index/columns):")
df_basic


📋 Basic DataFrame (auto index/columns):


,0,1,2,3
0,85,92,78,95
1,88,76,91,82
2,95,89,84,90


In [29]:
# Method 2: Custom row and column labels
df_labeled = pd.DataFrame(scores, index=students, columns=subjects)
print("\n Labeled DataFrame (custom index & columns):")
df_labeled


 Labeled DataFrame (custom index & columns):


,Math,Science,English,History
Alice,85,92,78,95
Bob,88,76,91,82
Carol,95,89,84,90


In [30]:
# Method 3: Add metadata (index/column names)
df_enhanced = df_labeled.copy()
df_enhanced.index.name = 'Student'
df_enhanced.columns.name = 'Subject'

print("\n✨ Enhanced with Metadata:")
print(f"\nIndex name: '{df_enhanced.index.name}'")
print(f"Column name: '{df_enhanced.columns.name}'")
df_enhanced


✨ Enhanced with Metadata:

Index name: 'Student'
Column name: 'Subject'


Subject,Math,Science,English,History
Student,,,,
Alice,85,92,78,95
Bob,88,76,91,82
Carol,95,89,84,90


####  Conversion Parameters & Options

| **Parameter** | **Purpose** | **Example** | **Default** |
|---------------|-------------|-------------|-------------|
| `data` | NumPy array to convert | `np.array([[1,2],[3,4]])` | Required |
| `columns` | Column labels | `['A', 'B']` | `[0, 1, 2, ...]` |
| `index` | Row labels | `['row1', 'row2']` | `[0, 1, 2, ...]` |
| `dtype` | Force specific data type | `dtype=float` | Inferred from array |
| `copy` | Create copy vs view | `copy=False` | `True` (safer) |

- **Set `copy=False`** for large arrays to save memory (if you won't modify the original)

**💡 Pro Tips**:- **Check data types** after conversion with `df.dtypes`

- **Always specify meaningful column names** for better code readability
- **Use `pd.Index()`** to add names to your index/columns for richer metadata

###  Pandas DataFrame → NumPy Array

**When to convert**: You need NumPy's computational speed and mathematical functions.

####  Conversion Methods & Performance

**Common scenarios**:

-  Performance-critical computations (linear algebra, statistics)
-  Integration with scientific computing libraries (SciPy, scikit-learn)
-  Machine learning algorithms that expect NumPy arrays
-  Mathematical operations not available in pandas

In [31]:
print(" PANDAS TO NUMPY CONVERSION")
print("=" * 40)

# Start with a DataFrame of tech stock prices
stock_data = pd.DataFrame(
    {
        "AAPL":  [150.0, 152.5, 148.2, 155.1],
        "GOOGL": [2800.0, 2750.0, 2825.0, 2900.0],
        "MSFT":  [300.0, 305.0, 298.5, 310.0],
        "TSLA":  [800.0, 795.0, 820.0, 815.0],
    },
    index=["Day 1", "Day 2", "Day 3", "Day 4"]
)

print("📈 Original DataFrame:")
print(stock_data)
print(f"\nShape: {stock_data.shape}")
print("Dtypes:")
print(stock_data.dtypes)
print(f"Memory usage (deep): {stock_data.memory_usage(deep=True).sum()} bytes")


 PANDAS TO NUMPY CONVERSION
📈 Original DataFrame:
        AAPL   GOOGL   MSFT   TSLA
Day 1  150.0  2800.0  300.0  800.0
Day 2  152.5  2750.0  305.0  795.0
Day 3  148.2  2825.0  298.5  820.0
Day 4  155.1  2900.0  310.0  815.0

Shape: (4, 4)
Dtypes:
AAPL     float64
GOOGL    float64
MSFT     float64
TSLA     float64
dtype: object
Memory usage (deep): 344 bytes


In [32]:
# -----------------------------
# Method 1: .to_numpy() (recommended)
# -----------------------------
print("\n🔢 Method 1 — .to_numpy() (recommended):")
array_modern = stock_data.to_numpy()
print(f"Shape: {array_modern.shape}")
print(f"Data type: {array_modern.dtype}")
array_modern



🔢 Method 1 — .to_numpy() (recommended):
Shape: (4, 4)
Data type: float64


array([[ 150. , 2800. ,  300. ,  800. ],
       [ 152.5, 2750. ,  305. ,  795. ],
       [ 148.2, 2825. ,  298.5,  820. ],
       [ 155.1, 2900. ,  310. ,  815. ]])

In [33]:
# -----------------------------
# Method 2: .values (legacy but still works)
# -----------------------------
print("\n⚠️  Method 2 — .values (legacy):")
array_legacy = stock_data.values
print(array_legacy)
print(f"Shape: {array_legacy.shape}")
print(f"Same result as .to_numpy()? {np.array_equal(array_modern, array_legacy)}")


⚠️  Method 2 — .values (legacy):
[[ 150.  2800.   300.   800. ]
 [ 152.5 2750.   305.   795. ]
 [ 148.2 2825.   298.5  820. ]
 [ 155.1 2900.   310.   815. ]]
Shape: (4, 4)
Same result as .to_numpy()? True


In [34]:
# -----------------------------
# Method 3: Force a specific dtype (e.g., int32)
# -----------------------------
print("\n Method 3 — Force data type (int32):")
array_int = stock_data.to_numpy(dtype=np.int32)
print(array_int)
print(f"Shape: {array_int.shape}")
print(f"Data type: {array_int.dtype}")
print(f"Preview (first 2 rows):\n{array_int[:2]}")


 Method 3 — Force data type (int32):
[[ 150 2800  300  800]
 [ 152 2750  305  795]
 [ 148 2825  298  820]
 [ 155 2900  310  815]]
Shape: (4, 4)
Data type: int32
Preview (first 2 rows):
[[ 150 2800  300  800]
 [ 152 2750  305  795]]


In [35]:
# -----------------------------
# Method 4: Convert specific columns only
# -----------------------------
print("\n🎯 Method 4 — Specific columns only (AAPL, MSFT):")
tech_stocks = stock_data[["AAPL", "MSFT"]].to_numpy()
print(tech_stocks)
print(f"Shape: {tech_stocks.shape}")
print(f"Data type: {tech_stocks.dtype}")


🎯 Method 4 — Specific columns only (AAPL, MSFT):
[[150.  300. ]
 [152.5 305. ]
 [148.2 298.5]
 [155.1 310. ]]
Shape: (4, 2)
Data type: float64


####  Method Comparison & Best Practices

| **Method**                | **Recommendation**           | **Pros**                                              | **Cons**                                              |
|---------------------------|------------------------------|-------------------------------------------------------|-------------------------------------------------------|
| `.to_numpy()`             | ✅ **Use by default**         | Clear intent, stable API, future-proof                | May copy data depending on dtypes/backing storage     |
| `.to_numpy(dtype=...)`    | ✅ **When you need a dtype**  | Explicit, consistent numeric types; easier math       | Casting may be lossy or fail; extra param to manage   |
| `.values`                 | ⚠️ **Legacy / avoid**         | Works on old pandas; quick                           | Ambiguous with mixed/extension dtypes; not future-proof |

**Best practices**
- **Convert only needed columns** to reduce memory and copying.
- **Specify `dtype`** when downstream computations require consistent numeric types.
- **Check for missing values** (`NaN`, `NaT`) before casting to integer dtypes.
- **Prefer `.to_numpy()` over `.values`** for clarity and forward compatibility.
- Be mindful that conversions may **return a copy**, not a view—modify the DataFrame directly if needed.


###  Preserving Metadata During Conversion

**Challenge:** Converting a DataFrame to a NumPy array drops **column names, index labels, and dtypes**.

####  Practical Strategies

**1) Save & restore index/columns (and dtypes)**

In [36]:
df = pd.DataFrame({
    "A": pd.Series([1, 2, 3], dtype="Int64"),
    "B": pd.Categorical(["x", "y", "x"]),
}, index=["r1", "r2", "r3"])

print("Original DataFrame:")
print(df)
print(f"\nOriginal dtype of A: {df['A'].dtype}")  # Int64

# Convert to NumPy
arr = df.to_numpy()
print(f"\nNumPy array:\n{arr}")
print(f"NumPy array dtype: {arr.dtype}")  # object!

# Save metadata
cols = df.columns.copy()
index = df.index.copy()
dtypes = df.dtypes.to_dict()

# Rebuild DataFrame
df_out = pd.DataFrame(arr, index=index, columns=cols)
print(f"\nRebuilt DataFrame:")
print(df_out)
print(f"\nRebuilt dtype of A: {df_out['A'].dtype}")  # object, values are <NA>

Original DataFrame:
       A  B
r1  <NA>  x
r2  <NA>  y
r3  <NA>  x

Original dtype of A: Int64

NumPy array:
[[<NA> 'x']
 [<NA> 'y']
 [<NA> 'x']]
NumPy array dtype: object

Rebuilt DataFrame:
       A  B
r1  <NA>  x
r2  <NA>  y
r3  <NA>  x

Rebuilt dtype of A: object


**2) Use a structured array to carry names**

In [37]:
# Keeps field names (and index) inside a NumPy structured array
rec = df.to_records(index=True)               # dtype has named fields

# Back to DataFrame later
df_back = pd.DataFrame.from_records(rec).set_index("index")
df_back

,A,B
index,,
r1,NaN,x
r2,NaN,y
r3,NaN,x


### 📄 Comprehensive Comparison: When to Use What

| **Aspect**           | **Pandas DataFrame**                                  | **NumPy Array**                               | **Best Use**                          |
|----------------------|--------------------------------------------------------|-----------------------------------------------|---------------------------------------|
| **Data labels**      | ✅ Rich metadata (column names, index, dtypes)         | ❌ No labels                                   | **DataFrame** for exploration         |
| **Memory usage**     | ❌ Higher overhead                                     | ✅ Minimal overhead                            | **Array** for large datasets          |
| **Speed**            | ❌ Slower (Python-level checks)                        | ✅ C-level, vectorized performance             | **Array** for heavy compute           |
| **Missing data**     | ✅ Robust (`NaN`, `NaT`, nullable dtypes)              | ⚠️ Limited (mainly float `NaN`)               | **DataFrame** for dirty data          |
| **Indexing**         | ✅ Label & integer (`.loc`, `.iloc`)                   | ❌ Integer/boolean only                        | **DataFrame** for complex queries     |
| **Broadcasting**     | ⚠️ Limited/indirect                                    | ✅ Full NumPy rules                            | **Array** for math ops                |
| **Data types**       | ✅ Mixed types per column                              | ❌ Homogeneous per array                       | **DataFrame** for mixed data          |
| **Plotting**         | ✅ Built-in `.plot()`                                  | ❌ Manual setup                                | **DataFrame** for quick viz           |
| **Linear algebra**   | ❌ Limited                                             | ✅ Rich (NumPy/BLAS/LAPACK)                    | **Array** for math/LA                 |
| **Group operations** | ✅ Powerful `.groupby()`, `.agg()`, `.transform()`     | ❌ Manual                                       | **DataFrame** for aggregations        |


#### 🧭 Recommended Workflow
- 🔍 **Exploratory analysis** → Use **Pandas**
- ⚡ **Heavy computation** → Convert to **NumPy** with `df.to_numpy()`
- 📊 **Results & visualization** → Convert back to **Pandas**
- 💾 **Export & storage** → Use **Pandas** (CSV/Parquet/SQL)

**Flow:** Pandas (explore) → NumPy (compute) → Pandas (analyze/plot) → Pandas (export)

> **Tip:** Convert only the needed columns and set `dtype` in `to_numpy(dtype=...)` when you need consistent numeric types. Handle missing values **before** casting.

##  Random Number Generation in NumPy

NumPy provides a variety of functions for generating random numbers through its numpy.random module. These functions can generate random numbers for different distributions and array shapes, making them essential for simulations, statistical sampling, and machine learning applications.

###  Key Functions for Random Number Generation in NumPy

#### `np.random.rand()`
It generates random numbers from a uniform distribution between 0 and 1.

In [38]:
# Generate a 2x3 array of random values between 0 and 1
rand_array = np.random.rand(2, 3)
print(rand_array)

[[0.39363552 0.47343566 0.85454739]
 [0.34000439 0.86964968 0.08813443]]


#### `np.random.randn()`

It generates random numbers from a standard normal distribution (mean = 0, standard deviation = 1). It is useful for simulating Gaussian distributed data

In [39]:
normal_array = np.random.randn(3, 3)
print(normal_array)

[[ 0.5376299   0.42818625 -0.2799933 ]
 [-1.27903074  0.51529432 -0.83428181]
 [ 2.18409673  0.57050721 -0.5806483 ]]


NumPy's [random](https://numpy.org/doc/stable/reference/random/index.html) module can be used to generate arrays of random numbers from several different probability distributions. For example, a `3x5` array of uniformly distributed random numbers can be generated using the `uniform` function of the `random` module.

#### `np.random.randint()`

It generate random integers within a specific range, you can specify `low` and  `high` values and the shape of the output array

In [40]:
# Generate a 4x4 matrix of random integers between 10 and 20
int_array = np.random.randint(10, 20, (4, 4))
print(int_array)

[[14 17 14 18]
 [15 17 19 17]
 [12 16 16 19]
 [15 16 11 18]]


#### `np.random.choice()`

Random selects elements from an input array

In [ ]:
# Select 5 random elements from the array [1, 2, 3, 4, 5] with replacement
choice_array = np.random.choice([1, 2, 3, 4, 5], size=5, replace=True)
print(choice_array)

[4 3 5 2 2]


#### `np.random.uniform()`
It generates random floating-point numbers between a specified range (low, high) that follow uniform distribution

In [ ]:
# Generate 6 random numbers from a normal distribution with mean=10 and std=2
custom_normal = np.random.normal(10, 2, size=6)
print(custom_normal)


[ 8.96337061  9.80411528  7.1623284   9.58883996 11.82330829 10.53508984]


In [ ]:
# Generate a 1D array of 5 random numbers between -5 and 5
uniform_array = np.random.uniform(-5, 5, size=5)
print(uniform_array)


[-4.76424391 -2.90776829  4.56871733  2.3230159  -4.72083461]


#### `np.random.normal()`

It generates random numbers from a normal distribution with a specified mean and standard deviation.

In [41]:
np.random.uniform(size = (3,5))

array([[0.82164157, 0.10137984, 0.79189132, 0.811572  , 0.482274  ],
       [0.05731031, 0.17458985, 0.70742628, 0.71177794, 0.89835528],
       [0.58268615, 0.12410737, 0.12141491, 0.72113236, 0.18378467]])

For a full list, please check the [offcial website](https://numpy.org/doc/stable/reference/random/generated/numpy.random.randint.html)

Random numbers can also be generated by Python's built-in [random](https://docs.python.org/3/library/random.html) module. However, it generates one random number at a time, which makes it much slower than NumPy's random module.

## Independent Study

### Practice exercise 2:
This exercise will show vectorized computations with NumPy. Vectorized computations help perform computations more efficiently, and also make the code concise.

**Q:** Read the (1) quantities of roll, bun, cake and bread required by 3 people - Ben, Barbara & Beth, from *food_quantity.csv*, (2) price of these food items in two shops - Target and Kroger, from *price.csv*. Find out which shop should each person go to minimize their expenses.

In [45]:
#Reading the datasets on food quantity and price
import pandas as pd
food_qty = pd.read_csv('./datasets/food_quantity.csv',index_col=0)
price = pd.read_csv('./datasets/price.csv',index_col=0)

In [ ]:
food_qty

,roll,bun,cake,bread
Person,,,,
Ben,6,5,3,1
Barbara,3,6,2,2
Beth,3,4,3,1


In [ ]:
price

,Target,Kroger
Item,,
roll,1.5,1.0
bun,2.0,2.5
cake,5.0,4.5
bread,16.0,17.0


First, let's start from a simple problem. We'll compute the expenses of Ben if he prefers to buy all food items from Target

In [46]:
%%time
# write a for loop to calculate the total cost of Ben's food if he shops at the Target store
total_cost = 0 
for food in food_qty.columns:

    total_cost += food_qty.loc['Ben',food]*price.loc[food,'Target']
total_cost

CPU times: total: 0 ns
Wall time: 1 ms


50.0

In [47]:
%%time
# using numpy 
total_cost = np.sum(food_qty.loc['Ben',]*price.loc[:,'Target'])
total_cost

CPU times: total: 0 ns
Wall time: 0 ns


50.0

Ben will spend $50 if he goes to Target

Now, let's add another layer of complication. We'll compute Ben's expenses for both stores - Target and Kroger

In [48]:
%%time
# using loops to calculate the total cost of food for Ben for all stores
total_cost = {}
for store in price.columns:
    total_cost[store] = 0
    for food in food_qty.columns:
        total_cost[store] += food_qty.loc['Ben',food]*price.loc[food,store]
total_cost

CPU times: total: 15.6 ms
Wall time: 999 μs


{'Target': 50.0, 'Kroger': 49.0}

In [49]:
%%time
# using numpy
total_cost = np.dot(food_qty.loc['Ben',],price.loc[:,:])
total_cost

CPU times: total: 15.6 ms
Wall time: 1 ms


array([50., 49.])

Ben will spend \$50 if he goes to Target, and \$49 if he goes to Kroger. Thus, he should choose Kroger.

Now, let's add the final layer of complication, and solve the problem. We'll compute everyone's expenses for both stores - Target and Kroger

In [50]:
%%timeit
store_expense = pd.DataFrame(0.0, columns=price.columns, index = food_qty.index)
for person in store_expense.index:
    for store in store_expense.columns:
        for food in food_qty.columns:
            store_expense.loc[person, store] += food_qty.loc[person, food]*price.loc[food, store]

store_expense

1.49 ms ± 22.5 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)


In [51]:
%%timeit
# using matrix multiplication in numpy
pd.DataFrame(np.dot(food_qty.values, price.values), columns=price.columns, index=food_qty.index)

11.3 μs ± 93.2 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)


Based on the above table, Ben should go to Kroger, Barbara to Target and Beth can go to either store.  

As the complexity of operations increases, the number of nested for-loops tends to grow, making the code more cumbersome and difficult to manage. In contrast, leveraging NumPy arrays allows for concise and straightforward implementation, regardless of the complexity. Vectorized computations are not only cleaner but also significantly faster, offering a more efficient solution. To take advantage of this, you first need to convert a pandas DataFrame to a NumPy array for matrix multiplication. Once the computation is complete, you can convert the results back to a pandas DataFrame for further analysis or display

### Practice exercise  3

Use matrix multiplication to find the average IMDB rating and average Rotten tomatoes rating for each genre - comedy, action, drama and horror. Use the data: *movies_cleaned.csv*. Which is the most preferred genre for IMDB users, and which is the least preferred genre for Rotten Tomatoes users?

**Hint:** 
1. Create two matrices - one containing the IMDB and Rotten Tomatoes ratings, and the other containing the genre flags (comedy/action/drama/horror). 

2. Multiply the two matrices created in 1.

3. Divide each row/column of the resulting matrix by a vector having the number of ratings in each genre to get the average rating for the genre.

**Solution:**

In [53]:
#| echo: false
data = pd.read_csv('./Datasets/movies_cleaned.csv')
data.head()

,Title,IMDB Rating,Rotten Tomatoes Rating,Running Time min,Release Date,US Gross,Worldwide Gross,Production Budget,comedy,Action,drama,horror
0,Broken Arrow,5.8,55,108,Feb 09 1996,70645997,148345997,65000000,0,1,0,0
1,Brazil,8.0,98,136,Dec 18 1985,9929135,9929135,15000000,1,0,0,0
2,The Cable Guy,5.8,52,95,Jun 14 1996,60240295,102825796,47000000,1,0,0,0
3,Chain Reaction,5.2,13,106,Aug 02 1996,21226204,60209334,55000000,0,1,0,0
4,Clash of the Titans,5.9,65,108,Jun 12 1981,30000000,30000000,15000000,0,1,0,0


In [54]:
#| echo: false
# Getting ratings of all movies
drating = data[['IMDB Rating','Rotten Tomatoes Rating']]
drating_num = drating.to_numpy() #Converting the data to NumPy array
drating_num

array([[ 5.8, 55. ],
       [ 8. , 98. ],
       [ 5.8, 52. ],
       ...,
       [ 7. , 65. ],
       [ 5.7, 26. ],
       [ 6.7, 82. ]])

In [55]:
#| echo: false
# Getting the matrix indicating the genre of all movies
dgenre = data.iloc[:,8:12]
dgenre_num = dgenre.to_numpy() #Converting the data to NumPy array
dgenre_num

array([[0, 1, 0, 0],
       [1, 0, 0, 0],
       [1, 0, 0, 0],
       ...,
       [1, 0, 0, 0],
       [0, 1, 0, 0],
       [0, 1, 0, 0]], dtype=int64)

We'll first find the total IMDB and Rotten tomatoes ratings for all movies of each genre, and then divide them by the number of movies of the corresponding genre to find the average rating for the genre.

For finding the total IMDB and Rotten tomatoes ratings, we'll multiply `drating_num` with `dgenre_num`. However, before multiplying, we'll check if their shapes are compatible for matrix multiplication.

In [56]:

#Shape of drating_num
drating_num.shape

(980, 2)

In [57]:
#Shape of dgenre_num
dgenre_num.shape

(980, 4)

Note that the above shapes are not compatible for matrix multiplication. We'll transpose `dgenre_num` to make the shapes compatible.

In [58]:
#Total IMDB and Rotten tomatoes ratings for each genre
ratings_sum_genre = drating_num.T.dot(dgenre_num)
ratings_sum_genre

array([[ 1785.6,  1673.1,  1630.3,   946.2],
       [14119. , 13725. , 14535. ,  6533. ]])

In [59]:
#Number of movies in the data will be stored in 'rows', and number of columns stored in 'cols'
rows, cols = data.shape

In [60]:
#Getting number of movies in each genre
movies_count_genre = dgenre_num.T.dot(np.ones(rows))
movies_count_genre

array([302., 264., 239., 154.])

In [61]:
#Finding the average IMDB and average Rotten tomatoes ratings for each genre
ratings_sum_genre/movies_count_genre

array([[ 5.91258278,  6.3375    ,  6.82133891,  6.14415584],
       [46.75165563, 51.98863636, 60.81589958, 42.42207792]])

In [62]:
pd.DataFrame(ratings_sum_genre/movies_count_genre,columns = ['comedy','Action','drama','horror'],
             index = ['IMDB Rating','Rotten Tomatoes Rating'])

,comedy,Action,drama,horror
IMDB Rating,5.912583,6.337500,6.821339,6.144156
Rotten Tomatoes Rating,46.751656,51.988636,60.815900,42.422078


IMDB users prefer *drama*, and are amused the least by *comedy* movies, on an average. However, Rotten tomatoes critics would rather watch *comedy* than *horror* movies, on an average.


### Practice exercise 4: 

**Random number generation using NumPy**

 Suppose 500 people eat at Food cart 1, and another 500 eat at Food cart 2, everyday.

The waiting time at Food cart 2 has a normal distribution with mean 8 minutes and standard deviation 3 minutes, while the waiting time at Food cart 1 has a uniform distribution with minimum 5 minutes and maximum 25 minutes. 

Simulate a dataset containing waiting times for 500 ppl for 30 days in each of the food joints. Assume that the waiting times are measured simultaneously at a certain time in both places, i.e., the observations are paired.

**On how many days is the average waiting time at Food cart 2 higher than that at Food cart 1?**

**What percentage of times the waiting time at Food cart 2 was higher than the waiting time at Food cart 1?**

Try both approaches: (1) Using loops to generate data, (2) numpy array to generate data. Compare the time taken in both approaches. 

In [64]:
import time as tm

In [65]:
#Method 1: Using loops
start_time = tm.time() #Current system time

#Initializing waiting times for 500 ppl over 30 days
waiting_times_FoodCart1 = pd.DataFrame(0,index=range(500),columns=range(30)) #FoodCart1
waiting_times_FoodCart2 = pd.DataFrame(0,index=range(500),columns=range(30)) #FoodCart2
import random as rm
for i in range(500):  #Iterating over 500 ppl
    for j in range(30): #Iterating over 30 days
        waiting_times_FoodCart2.iloc[i,j] = rm.gauss(8,3) #Simulating waiting time in FoodCart2 for the ith person on jth day
        waiting_times_FoodCart1.iloc[i,j] = rm.uniform(5,25) #Simulating waiting time in FoodCart1 for the ith person on jth day
time_diff = waiting_times_FoodCart2-waiting_times_FoodCart1

print("On ",sum(time_diff.mean()>0)," days, the average waiting time at FoodCart2 higher than that at FoodCart1")
print("Percentage of times waiting time at FoodCart2 was greater than that at FoodCart1 = ",100*(time_diff>0).sum().sum()/(30*500),"%")
end_time = tm.time() #Current system time
print("Time taken = ", end_time-start_time)

C:\Users\lsi8012\AppData\Local\Temp\ipykernel_28136\4118008220.py:10: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '3.8418066501543233' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.
  waiting_times_FoodCart2.iloc[i,j] = rm.gauss(8,3) #Simulating waiting time in FoodCart2 for the ith person on jth day
C:\Users\lsi8012\AppData\Local\Temp\ipykernel_28136\4118008220.py:11: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '19.485539894979468' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.
  waiting_times_FoodCart1.iloc[i,j] = rm.uniform(5,25) #Simulating waiting time in FoodCart1 for the ith person on jth day
C:\Users\lsi8012\AppData\Local\Temp\ipykernel_28136\4118008220.py:10: FutureWarning: Setting an item of incompatible dtype is deprecated and will

On  0  days, the average waiting time at FoodCart2 higher than that at FoodCart1
Percentage of times waiting time at FoodCart2 was greater than that at FoodCart1 =  16.38 %
Time taken =  3.045159339904785


In [66]:
#Method 2: Using NumPy arrays
start_time = tm.time()
waiting_time_FoodCart2 = np.random.normal(8,3,size = (500,30)) #Simultaneously generating the waiting times of 500 ppl over 30 days in FoodCart2
waiting_time_FoodCart1 = np.random.uniform(5,25,size = (500,30)) #Simultaneously generating the waiting times of 500 ppl over 30 days in FoodCart1
time_diff = waiting_time_FoodCart2-waiting_time_FoodCart1
print("On ",(time_diff.mean()>0).sum()," days, the average waiting time at FoodCart2 higher than that at FoodCart1")
print("Percentage of times waiting time at FoodCart2 was greater than that at FoodCart1 = ",100*(time_diff>0).sum()/15000,"%")
end_time = tm.time()
print("Time taken = ", end_time-start_time)

On  0  days, the average waiting time at FoodCart2 higher than that at FoodCart1
Percentage of times waiting time at FoodCart2 was greater than that at FoodCart1 =  16.26 %
Time taken =  0.0


The approach with NumPy is much faster than the one with loops.

### Practice exercise 5

**Bootstrapping:** Find the 95% confidence interval of mean profit for 'Action' movies, using Bootstrapping.

Bootstrapping is a non-parametric method for obtaining confidence interval. Use the algorithm below to find the confidence interval:

1. Find the profit for each of the 'Action' movies. Suppose there are *N* such movies. We will have a *Profit* column with *N* values. \
2. Randomly sample *N* values with replacement from the *Profit* column \
3. Find the mean of the *N* values obtained in (b) \
4. Repeat steps (b) and (c) *M=1000* times \
5. The 95% Confidence interval is the range between the 2.5% and 97.5% percentile values of the 1000 means obtained in (c) \
Use the *movies_cleaned.csv* dataset. 

**Solution:**